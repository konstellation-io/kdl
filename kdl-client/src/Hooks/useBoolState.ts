import { useState } from 'react';

// TODO: use this hook in more places
function useBoolState(defaultValue = false) {
  const [value, setValue] = useState(defaultValue);

  function toggle() {
    setValue(!value);
  }

  function activate() {
    setValue(true);
  }

  function deactivate() {
    setValue(false);
  }

  return {
    value,
    setValue,
    toggle,
    activate,
    deactivate,
  };
}

export default useBoolState;
