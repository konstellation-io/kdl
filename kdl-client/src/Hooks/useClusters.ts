import ROUTE, { buildRoute } from 'Constants/routes';
import { useEffect, useState } from 'react';

import { Action } from 'Pages/Clusters/components/Cluster/Cluster';
import IconStart from '@material-ui/icons/PlayArrow';
import IconStop from '@material-ui/icons/Stop';
import { LocalClusterStates } from 'Pages/Clusters/components/Cluster/LocalCluster';
import { RemoteClusterStates } from 'Pages/Clusters/components/Cluster/RemoteCluster';
import { ipcRenderer } from 'electron';

export enum ClusterType {
  LOCAL = 'local',
  REMOTE = 'remote',
}
export type Cluster = {
  id: string;
  name: string;
  type: ClusterType;
  state: LocalClusterStates | RemoteClusterStates;
  url?: string;
};

export type AddCluster = {
  name: string;
  type: ClusterType;
  url?: string;
};

function useClusters() {
  const [clusters, setClusters] = useState<Cluster[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    function onStoreUpdate(_: unknown, newClusters: Cluster[]) {
      setClusters(newClusters);
    }

    ipcRenderer.send('subscribeToValue', 'clusters');
    ipcRenderer.on('subscribeToValueReply', onStoreUpdate);

    async function fetchData() {
      const storeClusters = await ipcRenderer.invoke(
        'getStoreValue',
        'clusters'
      );
      setClusters(storeClusters);
      setLoading(false);
    }
    fetchData();

    return () => {
      ipcRenderer.removeListener('subscribeToValueReply', onStoreUpdate);
    };
  }, []);

  function getCluster(id: string): Cluster | undefined {
    return clusters.find((cluster) => cluster.id === id);
  }

  function getClusterActions(
    state: LocalClusterStates | RemoteClusterStates,
    id: string
  ) {
    let actions: Action[] = [];

    switch (state) {
      case RemoteClusterStates.SIGNED_OUT:
        actions = [
          {
            label: 'SIGN IN',
            Icon: IconStart,
            to: buildRoute.cluster(ROUTE.CLUSTER_LOGIN, id),
          },
        ];
        break;
      case RemoteClusterStates.SIGNED_IN:
        actions = [
          {
            label: 'SIGN OUT',
            Icon: IconStop,
            onClick: () => ipcRenderer.send('clusterLogout', id),
          },
        ];
        break;
      case LocalClusterStates.STARTED:
        actions = [
          {
            label: 'STOP',
            Icon: IconStop,
            onClick: () => ipcRenderer.send('stopLocalCluster', id),
          },
        ];
        break;
      case LocalClusterStates.STOPPED:
        actions = [
          {
            label: 'START',
            Icon: IconStart,
            onClick: () => ipcRenderer.send('startLocalCluster', id),
          },
        ];
        break;
      default:
        break;
    }

    return actions;
  }

  return {
    clusters,
    loading,
    getCluster,
    getClusterActions,
  };
}

export default useClusters;
