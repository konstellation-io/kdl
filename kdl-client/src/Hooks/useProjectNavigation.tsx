import { useMemo } from 'react';
import ROUTE, { buildRoute } from '../Constants/routes';
import IconHome from '@material-ui/icons/Dashboard';
import IconSettings from '@material-ui/icons/Settings';
import { OverridableComponent } from '@material-ui/core/OverridableComponent';
import { SvgIconTypeMap } from '@material-ui/core';

export interface RouteConfiguration {
  label: string;
  Icon: OverridableComponent<SvgIconTypeMap<{}, 'svg'>>;
}

export const projectRoutesConfiguration: {
  [key: string]: RouteConfiguration;
} = {
  [ROUTE.PROJECT_OVERVIEW]: {
    label: 'Overview',
    Icon: IconHome,
  },
  [ROUTE.PROJECT_TOOLS]: {
    label: 'Tools',
    Icon: IconSettings,
  },
};

export interface EnhancedRouteConfiguration extends RouteConfiguration {
  to: string;
}

function useProjectNavigation(serverId: string, projectId: string) {
  const routesConfigurations: EnhancedRouteConfiguration[] = useMemo(
    () =>
      Object.entries(projectRoutesConfiguration).map(
        ([routeString, { label, Icon }]) => {
          const route = routeString as ROUTE;
          return {
            to: buildRoute.project(route, serverId, projectId),
            label,
            Icon,
          };
        }
      ),
    [serverId, projectId]
  );
  return routesConfigurations;
}

export default useProjectNavigation;
