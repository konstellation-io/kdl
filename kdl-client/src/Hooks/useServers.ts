import { StoreKey, StoreUpdate } from 'types/store';
import { useEffect, useState } from 'react';

import { Action } from 'Pages/Servers/components/Server/Server';
import IconStart from '@material-ui/icons/PlayArrow';
import IconStop from '@material-ui/icons/Stop';
import { LocalServerStates } from 'Pages/Servers/components/Server/LocalServer';
import { RemoteServerStates } from 'Pages/Servers/components/Server/RemoteServer';
import { ipcRenderer } from 'electron';

export enum ServerType {
  LOCAL = 'local',
  REMOTE = 'remote',
}
export type Server = {
  id: string;
  name: string;
  type: ServerType;
  state: LocalServerStates | RemoteServerStates;
  url?: string;
};

export type AddServer = {
  name: string;
  type: ServerType;
  url?: string;
};

function useServers() {
  const [servers, setServers] = useState<Server[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    function onStoreUpdate(_: unknown, { key, value }: StoreUpdate) {
      if (key === StoreKey.SERVERS) {
        setServers(value);
      }
    }

    ipcRenderer.send('subscribeToValue', StoreKey.SERVERS);
    ipcRenderer.on('subscribeToValueReply', onStoreUpdate);

    async function fetchData() {
      const storeServers = await ipcRenderer.invoke(
        'getStoreValue',
        StoreKey.SERVERS
      );
      setServers(storeServers);
      setLoading(false);
    }
    fetchData();

    return () => {
      ipcRenderer.removeListener('subscribeToValueReply', onStoreUpdate);
    };
  }, []);

  function getServer(id: string): Server | undefined {
    return servers.find((server) => server.id === id);
  }

  function getServerActions(
    state: LocalServerStates | RemoteServerStates,
    id: string
  ) {
    let actions: Action[] = [];

    switch (state) {
      case LocalServerStates.STARTED:
        actions = [
          {
            label: 'STOP',
            Icon: IconStop,
            onClick: () => ipcRenderer.send('stopLocalServer', id),
          },
        ];
        break;
      case LocalServerStates.STOPPED:
        actions = [
          {
            label: 'START',
            Icon: IconStart,
            onClick: () => ipcRenderer.send('startLocalServer', id),
          },
        ];
        break;
      default:
        break;
    }

    return actions;
  }

  return {
    servers,
    loading,
    getServer,
    getServerActions,
  };
}

export default useServers;
