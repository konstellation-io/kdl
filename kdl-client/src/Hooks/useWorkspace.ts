import { StoreKey, StoreUpdate } from 'types/store';
import { useEffect, useState } from 'react';

import { ipcRenderer } from 'electron';

export type Workspace = {
  project: {
    navigationOpened: boolean;
  };
};

function useWorkspace() {
  const [workspace, setWorkspace] = useState<Workspace>();
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    function onStoreUpdate(_: unknown, { key, value }: StoreUpdate) {
      if (key === StoreKey.WORKSPACE) {
        setWorkspace(value);
      }
    }

    ipcRenderer.send('subscribeToValue', StoreKey.WORKSPACE);
    ipcRenderer.on('subscribeToValueReply', onStoreUpdate);

    async function fetchData() {
      const storeWorkspace = await ipcRenderer.invoke(
        'getStoreValue',
        StoreKey.WORKSPACE
      );
      setWorkspace(storeWorkspace);
      setLoading(false);
    }
    fetchData();

    return () => {
      ipcRenderer.removeListener('subscribeToValueReply', onStoreUpdate);
    };
  }, []);

  function toggleProjectNavOpened() {
    if (workspace) {
      ipcRenderer.send('setStoreValue', {
        key: 'workspace.project.navigationOpened',
        value: !workspace.project.navigationOpened,
      });
    }
  }

  return {
    workspace,
    toggleProjectNavOpened,
    loading,
  };
}

export default useWorkspace;
