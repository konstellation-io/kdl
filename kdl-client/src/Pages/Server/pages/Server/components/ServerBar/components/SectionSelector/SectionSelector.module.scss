@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  z-index: 3;
  ul {
    list-style-type: none;
    padding: 0;
    margin: 0;
    .section {
      @include font-body;
      color: font-color(lowlight);
      padding: $grid-unit;
      &.selectedSection {
        background-color: palette(highlight);
        color: font-color(grey);
      }
    }
  }
}
