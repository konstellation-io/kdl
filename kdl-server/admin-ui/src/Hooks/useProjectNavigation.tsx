import ROUTE, { buildRoute } from '../Constants/routes';

import IconHome from '@material-ui/icons/Dashboard';
import IconKG from '@material-ui/icons/EmojiObjects';
import IconSettings from '@material-ui/icons/Settings';
import { OverridableComponent } from '@material-ui/core/OverridableComponent';
import { SvgIconTypeMap } from '@material-ui/core';
import { useMemo } from 'react';

export interface RouteConfiguration {
  id: string;
  label: string;
  Icon: OverridableComponent<SvgIconTypeMap<{}, 'svg'>>;
}

export const projectRoutesConfiguration: {
  [key: string]: RouteConfiguration;
} = {
  [ROUTE.PROJECT_OVERVIEW]: {
    id: 'overview',
    label: 'Overview',
    Icon: IconHome,
  },
  [ROUTE.PROJECT_TOOLS]: {
    id: 'tools',
    label: 'Tools',
    Icon: IconSettings,
  },
  [ROUTE.PROJECT_KG]: {
    id: 'knowledge-graph',
    label: 'Knowledge Graph',
    Icon: IconKG,
  },
};

export interface EnhancedRouteConfiguration extends RouteConfiguration {
  to: string;
}

function useProjectNavigation(serverId: string, projectId: string) {
  const routesConfigurations: EnhancedRouteConfiguration[] = useMemo(
    () =>
      Object.entries(projectRoutesConfiguration).map(
        ([routeString, { id, label, Icon }]) => ({
          to: buildRoute.project(routeString as ROUTE, serverId, projectId),
          id,
          label,
          Icon,
        })
      ),
    [serverId, projectId]
  );
  return routesConfigurations;
}

export default useProjectNavigation;
