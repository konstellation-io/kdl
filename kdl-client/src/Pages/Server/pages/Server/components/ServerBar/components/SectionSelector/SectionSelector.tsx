import React, { FC } from 'react';
import styles from './SectionSelector.module.scss';
import { Link } from 'react-router-dom';
import cx from 'classnames';
import { enhancedRouteConfiguration } from 'Hooks/useProjectNavigation';

type Props = {
  options: enhancedRouteConfiguration[];
  selectedSection: string;
};
const SectionSelector: FC<Props> = ({ options, selectedSection }) => (
  <div className={styles.container}>
    <ul>
      {options.map((option: enhancedRouteConfiguration) => {
        const Icon = option.icon;
        return (
          <Link to={option.to}>
            <li
              className={cx(styles.section, {
                [styles.selectedSection]:
                  option.label.toLowerCase() === selectedSection.toLowerCase(),
              })}
              key={option.id}
            >
              <Icon className={cx('icon-regular', styles.icon)} />
              <span className={styles.label}>{option.label}</span>
            </li>
          </Link>
        );
      })}
    </ul>
  </div>
);
export default SectionSelector;
