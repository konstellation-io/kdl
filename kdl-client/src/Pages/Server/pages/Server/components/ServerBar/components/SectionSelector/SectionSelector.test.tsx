import SectionSelector from './SectionSelector';
import React from 'react';
import { shallow } from 'enzyme';
import { EnhancedRouteConfiguration } from '../../../../../../../../Hooks/useProjectNavigation';
import IconHome from '@material-ui/icons/Dashboard';

const section: EnhancedRouteConfiguration = {
  Icon: IconHome,
  label: 'foo',
  to: 'bar',
};

const props = {
  options: [section],
  selectedSection: 'baz',
};

let component;
beforeEach(() => {
  component = shallow(<SectionSelector {...props} />);
});

describe('SectionSelector component', () => {
  it('should render without crashing', function () {
    expect(component).toMatchSnapshot();
  });
});
