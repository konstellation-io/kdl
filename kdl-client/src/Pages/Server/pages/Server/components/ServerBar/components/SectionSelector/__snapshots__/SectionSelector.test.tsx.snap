// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionSelector component should render without crashing 1`] = `
<div
  className="container"
>
  <ul>
    <NavLink
      activeClassName="selectedSection"
      className="section"
      key="foo"
      onClick={[MockFunction]}
      to="bar"
    >
      <NavigationButton
        Icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "DashboardIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        iconSize="icon-small"
        label="foo"
      />
    </NavLink>
  </ul>
</div>
`;
